### ISSUE TITLE
Infinite length ranges inconsistently implemented

### ISSUE THREAD

**Issue by dlfivefifty** [JuliaLang/julia#15562]
The following is inconsistent:

```julia
k=1:Inf
first(k)   # works
last(k)   # works
k[1]       # throws inexact error
```

In 0.5, show is also broken.  

I would argue that infinite lengths are worth implementing: it is more convenient to do

```julia
for k=1:Inf
   # do stuff and call break
end
```

then

```julia
k=1
while true
  # do stuff and call break
end
```

**Comment by nalimilan** [comment]
Looks like `checkbounds` doesn't like `Inf`. Cc: @mbauman

I think the error when trying to print the range is caused by `isempty` failing, which itself comes from `length` not working. And `length` cannot work for such ranges since `Inf` cannot be represented by an integer type.

Also, the `for` patterns you indicate are unlikely to be a good idea. `Inf` only exists for floating point types, so `1:Inf` is a float range, while indices are (generally) integers. But you can do `for x in itr` and break when you want.

**Comment by mbauman** [comment]
Yeah, the root trouble here is [`length`](https://github.com/JuliaLang/julia/blob/05bdbb0f7608e259b646339d4f5d672baaf1563c/base/range.jl#L324).  You could try returning a floating point length, but I have a feeling that will cause lots of troubles elsewhere… we have a pretty well-entrenched assumption that lengths are integers.

In this case, you can use `for i in countfrom(1)`.  A mathconst-like infinity would allow `colon(x::Int, ::Infinity) = countfrom(x)`, but adding that creates more difficulties than it solves: https://github.com/JuliaLang/julia/pull/6062.

**Comment by nalimilan** [comment]
Maybe `1:Inf` should raise an error directly, and point users to `countfrom`?

**Comment by dlfivefifty** [comment]
I think allowing length to return a float is indeed troublesome.  

Reading #6062 convinced me that I should create a `MathConst{:∞}` in my package:  I have a data structure that represents infinite-dimensional arrays, and need to support
```julia
A[2:∞,2:∞]
```
to take subarrays.  I can then overload `colon(x::Int,::MathConst{:∞})` to return `countfrom(x)`.

**Comment by eschnett** [comment]
You could also overload `colon(x::Int, y::Float64)`, and add a run-time check whether `y` is `Inf` or `-Inf`.

Another option (that I took in [https://github.com/eschnett/FlexibleArrays.jl](url)) is to use a different syntax for this case, i.e. to not use the `colon` operator here. This has the advantage that it is easier to get correct, as you don't have to shoehorn your handling of infinity into Base's notion of a range. (What if other packages did the same -- would that lead to a conflict if one uses both packages?)

For example, you could define a type `StartFrom` that represents an open interval, and a function `startfrom(x::Int)` that returns such a type. While overloading `getindex`, you would then handle this type as well.

**Comment by dlfivefifty** [comment]
overriding `colon(x::Int,y::Float64)` is a bad idea: it may affect things outside the package.  Creating a new type representing ∞ avoids this issue, as the type is package specific: No one is going to accidentally call `1:MyPackage.∞`

**Comment by dlfivefifty** [comment]
For my own usage, I've done the following to make infinite-length ranges work as expected:

```julia
## New Inf

const ∞ = Irrational{:∞}()

Base.show(io::IO, x::Irrational{:∞}) = print(io, "∞")
Base.convert{F<:AbstractFloat}(::Type{F},::Irrational{:∞}) = convert(F,Inf)

## My Count
# Need to reimplement, to avoid changing behaviour in Base

immutable Count{S<:Number}
    start::S
    step::S
end
countfrom(start::Number, step::Number) = Count(promote(start, step)...)
countfrom(start::Number)               = Count(start, one(start))
countfrom()                            = Count(1, 1)

Base.eltype{S}(::Type{Count{S}}) = S

Base.start(it::Count) = it.start
Base.next(it::Count, state) = (state, state + it.step)
Base.done(it::Count, state) = false

Base.length(it::Count) = ∞
getindex(it::Count,k) = it.start + it.step*(k-1)


## Colon

Base.colon(a::Real,b::Irrational{:∞}) = countfrom(a)
Base.colon(::Irrational{:∞},::AbstractFloat,::Irrational{:∞}) = [∞]
Base.colon(a::Real,st::Real,b::Irrational{:∞}) = countfrom(a,st)

Base.isinf(::Irrational{:∞}) = true
```

**Comment by mbauman** [comment]
`k=1.0:Inf` now throws an error upon construction.  I think doing this through a custom type is the way to go here (indeed: https://github.com/JuliaApproximation/InfiniteArrays.jl).  Are there any other changes to base that you need to satisfy that case?

**Comment by dlfivefifty** [comment]
So far it works with very few modifications. The only modifications that might be worth changing Base I can think of are:

1. The use of `OneTo`: I had to [override](https://github.com/JuliaApproximation/InfiniteArrays.jl/blob/5d5860dcb2eb1dcfd46517b3c050abd5235) `OneTo(::Infinity)` to return a special type `OneToInf`: a25b6/src/infrange.jl#L344  This is misleading as `OneTo(x)` should always return a type of `OneTo`. The proposed change to Base would be to change the function name to `oneto`.
2. Rewrite `vcat` and `hcat` in a broadcast-like manner with traits, etc. Lazy broadcasting works [amazing](https://github.com/JuliaApproximation/InfiniteArrays.jl/blob/master/src/lazybroadcasting.jl), though I haven't tried it since the broadcast rewrite. [Lazy concatination](https://github.com/JuliaApproximation/InfiniteArrays.jl/blob/master/src/lazyconcat.jl) needs the user to call `Vcat` or `Hcat`. 
3. `ReshapedArray` should work with non-`Int` lengths

Surprisingly, that's about it. And the package is really fun:
```julia
julia> using InfiniteArrays

julia> Vcat(Ones(1,∞), Hcat(Zeros(∞), Diagonal(1:∞)))
Vcat{Float64,2,Tuple{Ones{Float64,2,Tuple{Int64,InfiniteArrays.Infinity}},Hcat{Float64,Tuple{Zeros{Float64,1,Tuple{InfiniteArrays.Infinity}},Diagonal{Int64,InfiniteArrays.InfUnitRange{Int64}}}}}} with indices OneToInf()×OneToInf():
 1.0  1.0  1.0  1.0  1.0  1.0  1.0  …  
 0.0  1.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  2.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  3.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  4.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  5.0  0.0  …  
 0.0  0.0  0.0  0.0  0.0  0.0  6.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 0.0  0.0  0.0  0.0  0.0  0.0  0.0     
 ⋮                        ⋮         ⋱  
```

**Comment by dlfivefifty** [comment]
Also, I had to make `Infinity <: Integer`. This seemed like a bad idea at first, but I've come around to it: after all, `Inf isa Real`, and both `Infinity` and (positive) `Integer`s are cardinalities.

**Comment by mbauman** [comment]
That's great. In your bulleted list, 3 is #17372, 2 is #2326, and 1 is interesting.  Do you recall what the call-site(s) were that required that method?  Or was it just you calling `OneTo(∞)` directly?

**Comment by dlfivefifty** [comment]
An alternative (but probably overkill) solution is change `OneTo` so that `stop` doesn't need to have the same type:
```julia
struct OneTo{T<:Integer,V<:Integer} <: AbstractUnitRange{T}
    stop::T
    OneTo{T,V}(stop) where {T<:Integer,V<:Integer} = new(max(zero(T), stop))
end
OneTo{T}(stop::V) where V = OneTo{T,V}(stop)
OneTo(stop) = OneTo{typeof(one(T))}(stop)
```

**Comment by adienes** [comment]
infinite length ranges can no longer be constructed so this seems closeable
